spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/SparkMLTrainer.java

```diff
@@ -0,0 +1,145 @@
+   */
+  protected static class MLTrainerConfig extends PluginConfig {
+
+    @Description("The name of the FileSet to save the model to.")
+    protected String fileSetName;
+
+    @Description("Path of the FileSet to save the model to.")
+    protected String path;
+
+    @Nullable
+    @Description("A comma-separated sequence of fields that needs to be used for training.")
+    protected String featureFieldsToInclude;
+
+    @Nullable
+    @Description("A comma-separated sequence of fields that needs to be excluded from being used in training.")
+    protected String featureFieldsToExclude;
+
+    @Description("The field from which to get the label. It must be of type double.")
+    protected String labelField;
+
+    protected MLTrainerConfig() {
+      System.out.print("hello");
+    }
+
+    protected MLTrainerConfig(String fileSetName, String path, @Nullable String featureFieldsToInclude,
+                              @Nullable String featureFieldsToExclude, String labelField) {
+      this.fileSetName = fileSetName;
+      this.path = path;
+      this.featureFieldsToInclude = featureFieldsToInclude;
+      this.featureFieldsToExclude = featureFieldsToExclude;
+      this.labelField = labelField;
+    }
+
+    public void validate(Schema inputSchema) {
+      SparkUtils.validateConfigParameters(inputSchema, featureFieldsToInclude, featureFieldsToExclude, labelField,
+                                          null);
+      SparkUtils.validateLabelFieldForTrainer(inputSchema, labelField);
+    }
+  }
+
+  public SparkMLTrainer(MLTrainerConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    pipelineConfigurer.createDataset(config.fileSetName, FileSet.class);
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    Preconditions.checkArgument(inputSchema != null, "Input Schema must be a known constant.");
+    config.validate(inputSchema);
+  }
+
+  @Override
+  public void prepareRun(SparkPluginContext context) throws Exception { }
+
+  @Override
+  public void run(SparkExecutionPluginContext context, JavaRDD<StructuredRecord> input) throws Exception {
+    if (!input.isEmpty()) {
+      final Schema inputSchema = input.first().getSchema();
+      final Map<String, Integer> featuresList = SparkUtils.getFeatureList(inputSchema, config.featureFieldsToInclude,
+                                                                          config.featureFieldsToExclude,
+                                                                          config.labelField);
+      JavaRDD<LabeledPoint> trainingData = input.map(new Function<StructuredRecord, LabeledPoint>() {
+        @Override
+        public LabeledPoint call(StructuredRecord record) throws Exception {
+          List<Double> featureList = new ArrayList<>();
+          List<Integer> featureIndex = new ArrayList<>();
+          int counter = 0;
```

Comment:
Yes, in that way we can save two traversals. But the size of featureList and featureIndex could not be defined using featuresList.size(). It's because a record may have NULL values from the fields contained in featuresList. So if we create an array of  size 5 for 5 features, and if one of the input records have 2 features then unoccupied elements will be treated as 0 in case of integers. We are using Sparse Vectors in which generally 0's are avoided.

Please suggest for which approach should we go for.